File: easy-solution/galaxy4276/20.rs
/*
  Discuss 도움 받아서 해결함
*/

impl Solution {
  pub fn is_valid(s: String) -> bool {
  let mut stack  = vec![];
  for c in s.chars() {
      match c {
          ')' => match stack.pop() {
              Some(c) => {
                  if c != '(' {
                      return false;
                  }
              }
              None => return false,
          },
          ']' => match stack.pop() {
              Some(c) => {
                  if c != '[' {
                      return false;
                  }
              },
              None => return false,
          },
          '}' => match stack.pop() {
              Some(c) => {
                  if c != '{' {
                      return false;
                  }
              },
              None => return false,
          },
          _ => {
              stack.push(c);
          },
      }
  }
  return stack.is_empty();
  }
}

File: easy-solution/Antidote21/20.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> t; 
        //int i; 
        
        for(auto i:s) 
        {
            if(i == '(' || i =='{' || i == '[')
            {
                t.push(i);
            }
            else
            {
                // check condition for false
                if(t.empty() || (t.top() == '(' && i != ')') || (t.top() == '{' && i != '}') || (t.top() == '[' && i != ']'))
                {
                    return false;
                }
                
                t.pop(); // else pop from stack
            }
        }
        
        return t.empty(); // if stack is empty then it is valid, otherwise no
    }
};
